### PR TITLE
Add `advanced_parquet_index.rs` example of index in into parquet files

### PR DESCRIPTION
**Pull request by alamb**
## Which issue does this PR close?

Closes https://github.com/apache/datafusion/issues/10580

## Rationale for this change
See https://github.com/apache/datafusion/issues/10580

This shows how to use the APIs in  https://github.com/apache/datafusion/issues/9929

Building and using external indexes in DataFusion is an important feature. Adding an example of how to do so will help drive the design and APIs and help other people discover it more easiy

Specifically, this example illustrates how to configure the ParquetExec to read  only some row groups / pages *within* a file and shows how to avoid reading the metadata each time. This is required for low latency parquet query access

## What changes are included in this PR?

Add a new example: `parquet_index_advanced`

## Are there any user-facing changes?
This is an example

### CODE REVIEW

File: datafusion/common/src/column.rs
Comment (alamb) [inline datafusion/common/src/column.rs:130]: I added some small access APIs to make the example easier to read. I can revert these or put them in a different PR if reviewers prefer

**Review by hengfeiyang — APPROVED** [review]

File: datafusion-examples/examples/advanced_parquet_index.rs
Comment (adriangb) [inline datafusion-examples/examples/advanced_parquet_index.rs:177]: Seems like this is only taking an object store now and creating the directory internally

Comment (alamb) [reply datafusion-examples/examples/advanced_parquet_index.rs:177]: good call -- fixed in 79e1476a2

File: datafusion-examples/README.md
Comment (Weijun-H) [inline datafusion-examples/README.md:48]:
```suggestion
- [`advanced_parquet_index.rs`](examples/advanced_parquet_index.rs): Creates a detailed secondary index that covers the contents of several parquet files
```

File: datafusion-examples/examples/advanced_parquet_index.rs
Comment (Weijun-H) [inline datafusion-examples/examples/advanced_parquet_index.rs:64]:
```suggestion
/// This example demonstrates using low level DataFusion APIs to read only
```

File: datafusion-examples/examples/advanced_parquet_index.rs
Comment (Weijun-H) [inline datafusion-examples/examples/advanced_parquet_index.rs:84]:
```suggestion
/// 2. Use [`PruningPredicate`] for predicate analysis
```

File: datafusion-examples/examples/advanced_parquet_index.rs
Comment (Weijun-H) [inline datafusion-examples/examples/advanced_parquet_index.rs:263]:
```suggestion
            // and the reader is configured to read it, add a row selection
```

File: datafusion-examples/examples/advanced_parquet_index.rs
Comment (Weijun-H) [inline datafusion-examples/examples/advanced_parquet_index.rs:102]:
```suggestion
    // Create a custom table provider with our special index.
```

**Review by Weijun-H — APPROVED** [review]
Nice example! Thanks @alamb 👍

File: datafusion-examples/examples/advanced_parquet_index.rs
Comment (alamb) [inline datafusion-examples/examples/advanced_parquet_index.rs:94]: I added a diagram here to try and explain what is going on visually for those so inclined

Comment (adriangb) [reply datafusion-examples/examples/advanced_parquet_index.rs:94]: Just curious, do you generate these by hand? I was making them by hand with auto complete assistance, but wondering if there's a better tool.

Comment (alamb) [reply datafusion-examples/examples/advanced_parquet_index.rs:94]: I use this https://monodraw.helftone.com/

 I think there are other online ascii art diagrammers, but that monodraw works well for me

**Comment by alamb** [comment]
This PR is just waiting for a committer to approve it so I can merge it in. @crepererum  or @tustvold  might you have a moment to do so?

**Comment by alamb** [comment]
@crepererum  do you by any chance have a few minutes to review and approve this PR?

**Comment by alamb** [comment]
I can't merge this PR until a committer approves it:

Screenshot 2024-06-21 at 12 47 52 PM

@Dandandan or @thinkharderdev  any chance you have some time to review an example (mostly comments)?

File: datafusion-examples/examples/advanced_parquet_index.rs
Comment (thinkharderdev) [inline datafusion-examples/examples/advanced_parquet_index.rs:131]: Maybe add a note here that this can only happen if the parquet file has a `PageIndex`?

**Review by thinkharderdev — APPROVED** [review]
very nice!

File: datafusion-examples/examples/advanced_parquet_index.rs
Comment (alamb) [reply datafusion-examples/examples/advanced_parquet_index.rs:132]: Good call -- done in d2f64777a

**Comment by alamb** [comment]
Thank you everyone for the great comments and feedback

**Pull request merged**
